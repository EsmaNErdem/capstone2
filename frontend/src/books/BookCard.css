.BookCard {
    width: 100%;
    max-width: 200px;
    margin: 10px;
    padding: 10px;
    border: 1px solid #ddd;
    border-radius: 8px;
    overflow: hidden;
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
    transition: transform 0.3s ease-in-out;
  
    &:hover {
      transform: scale(1.02);
    }
  }
  
  .CardImage img {
    width: 100%;
    height: 150px; 
    object-fit: cover;
  }
  
  .CardContent {
    padding: 16px;
  }

  .Title {
    text-decoration: none;
  }

  h2 {
    color: #a074f3; 
    font-size: 1.2rem;
    margin-bottom: 8px;
  }

  h2:hover {
    color: #ff6347; 
    transition: color 0.3s ease;
  }
  
  h3 {
    font-size: .85rem;
    color: gray;
    margin-bottom: 12px;
    font-style: italic;
  }
  
  p {
    font-size: 1rem;
    margin-bottom: 12px;
  }

  .CardFooter {
    display: flex;
    justify-content: space-evenly;
    align-items: center;
    margin-top: 10px; 
  }

  .CardImage {
    border-radius: 8px;
  }
  
  @media (min-width: 768px) {
    .BookCard {
      max-width: 100%;
    }
  
    .CardImage img {
      height: 200px;
    }
  }
  
  @media (min-width: 992px) {
    .BookCard {
      max-width: calc(20%); /* Four books in a row with margin */
    }
  }
  
