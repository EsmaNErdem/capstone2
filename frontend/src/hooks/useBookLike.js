import { useState, useEffect, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user book like
 * This hook provides functionality to manage user likes on books, including retrieving initial like status and count,
 * updating the like status and count, and handling user interactions to like or unlike a book.
 * 
 * -Utilizes useContext data to send user like to database
 * -Creates state to update book like count and user book like status
 * -Use useEffect to avoid to recalculations
 * 
 * Returns user like status, book like count states and handleLike function
 */
const useBookLike = (bookId, initialLikes, bookData={}) => {
    console.debug("useBookLike"); 

    const { hasLikedBook, likeBook } = useContext(UserContext);
    const [liked, setLiked] = useState();
    const [likes, setLikes] = useState();
    const [error, setError] = useState(null);

    // Helper function to convert "undefined" or "null" strings to actual undefined
    const convertUndefined = (value) => (value === "undefined" || value === null ? undefined : value);

    // Convert undefined/null strings in the bookData object
    const convertedBookData = Object.fromEntries(
        Object.keys(bookData).map((key) => [key, convertUndefined(bookData[key])])
    );
        
    /**By using the useEffect, the liked status is only recalculated when the id avoiding unnecessary recalculations on every render.  */
    useEffect(() => {
      setLiked(hasLikedBook(bookId));
      setLikes(initialLikes)
    }, [bookId, initialLikes]);
    
    /**
     * Send user like to database
     * Updates the like count and status based on user actions.
     */ 
    const handleLikeBook = async () => {
        try {
            setError(null);
            if (liked) {
                const unlikedBookId = await likeBook(bookId);
                if (unlikedBookId === bookId) {
                    setLikes((likes) => likes - 1);
                    setLiked(false);
                }
            } else {
                const likedBookId = await likeBook(bookId, convertedBookData);
                if (likedBookId === bookId) {
                    setLikes((likes) => likes + 1);
                    setLiked(true);
                }
            }
        } catch (error) {
            setError('Error liking book.');
            console.error('Error handling book like:', error);
        }
    };
    
  return { liked, likes, error, handleLikeBook };
};

export default useBookLike;