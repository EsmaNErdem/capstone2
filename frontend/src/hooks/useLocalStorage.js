import { useState, useEffect } from "react";

/**
 * Custom Hook to save user info like token and username into localStorage
 * 
 * Creates a user state and looks in localStorage for current user info, if cannot find it, assign null
 * When user info updates, useEffect either cleans it or updates localStorage
 * 
 * Retunrs [user, setUser]
 */

const useLocalStorage = (localStorageKey, initialValue = null) => {
    let initialUserValue
    if(localStorage.getItem(localStorageKey)) {
        try {
            initialUserValue = JSON.parse(localStorage.getItem(localStorageKey))
        } catch (e) {
            console.error("useLocalStorage custom hook: problem loading localStorage data", e);
        }
    } else {
        initialUserValue = initialValue
    }

    const [user, setUser] = useState(initialUserValue)

    useEffect(() => {
        if(user === null) {
            localStorage.removeItem(localStorageKey)
        } else {
            localStorage.setItem(localStorageKey, JSON.stringify(user))
        }
    }, [localStorageKey, user])

    return [user, setUser]
}

export default useLocalStorage;