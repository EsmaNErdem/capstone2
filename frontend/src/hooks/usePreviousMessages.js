import { useState, useEffect } from "react";
import BookClubApi from "../api";

/** Custom Hook for fetching previous messages between receiver and current user
 * This hook makes API call to backend to retrieve related data and set messaegs state
 * 
 * Returns error and messages
 */
const usePreviousMessages = (receiver, currentUserName, websocket, isOpen) => {
    const [error, setError] = useState(null);
    const [messages, setMessages] = useState([]);

    /**
     * Fetches the previous messages between users from backend sends to child component after WebSocket connection is established
     */ 
    useEffect(function getPreviousMessages() {
        console.debug("usePreviousMessages hook useEffect getPreviousMessages");
    
        const getMessages = async () => {
            try{
                if (isOpen && websocket) {
                    const roomName = `${receiver},${currentUserName}`
                    const previousMessages = await BookClubApi.getRoomPreviousMessages(roomName);
                    setMessages([...previousMessages]);                 
                }
            } catch (e) {
                console.error("usePreviousMessages useEffect API call data loading error:", e);
                setError("An error occurred while fetching previous messages.");
            }
        }
        getMessages();
    }, [isOpen, receiver, websocket]);

    return { error, messages }
}

export default usePreviousMessages;