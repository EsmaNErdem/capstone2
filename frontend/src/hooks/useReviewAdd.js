import { useState, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user review submission on books/:id route components
 * This hook provides functionality to manage user review submission and updateting bookReviews state
 * 
 * -Utilizes useContext data to send user review to database
 * -Updates book review
 * 
 * Returns error and addBookReview function
 */
const useReviewAdd = (setBookReviews, bookData={}, errorHandlingState) => {
    console.debug("useReviewAdd"); 

    const { addUserReview } = useContext(UserContext);
    const [error, setError] = useState(null);

    // add current user's book review, function is called in BookReviewDrawer
    const addBookReview = async ({ review }) => {
        try {
            console.log({book: bookData, review})
            const newReview = await addUserReview({book: bookData, review});
            setBookReviews(r => [{...newReview, id:bookData.id}, ...r])
            return newReview
        } catch (error) {
            errorHandlingState(false)
            setError("Error adding book review.")
            console.error("Error adding book review:", error);
        }
    } 
    
    return { error, addBookReview }

}

export default useReviewAdd