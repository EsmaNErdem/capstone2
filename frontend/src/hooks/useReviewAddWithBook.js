import { useState, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user review submission for /reviews route components
 * This hook provides functionality to manage user review submission and updateting bookReviews state
 * 
 * -Utilizes useContext data to send user review to database
 * -Updates book review
 * 
 * Returns error and addBookReview function
 */
const useReviewAddWithBook = (setBookReviews, errorHandlingState) => {
    console.debug("useReviewAddWithBook"); 

    const { addUserReview } = useContext(UserContext);
    const [error, setError] = useState(null);

    // add current user's book review, function is called in BookReviewDrawer
    const addBookReview = async (review, bookData={}) => {
        try {
            const convertedBookData = convertNullInObject(bookData);
            const newReview = await addUserReview({book: convertedBookData, review});
            setBookReviews(r => [{...newReview, book_id:bookData.id, title:bookData.title, author:bookData.author, cover:bookData.cover, category:bookData.category, reviewLikeCount:0}, ...r])
            return newReview
        } catch (error) {
            errorHandlingState(false)
            setError("Error adding book review.")
            console.error("Error adding book review:", error);
        }
    } 
    
    return { error, addBookReview }

}

export default useReviewAddWithBook

const convertNullInObject = (obj) => {
    const convertedObj = {};
    for (const key in obj) {
      convertedObj[key] = obj[key] === null ? "" : obj[key];
    }
    return convertedObj;
  };