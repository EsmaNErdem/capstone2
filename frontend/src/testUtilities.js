import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  firstNsme: "testfirst",
  lastName: "testlast",
  email: "test@test.net",
  img: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => false, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const CurrUserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => true, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const NonUserProvider =
    ({ children, currentUser = null, hasLikedReview = () => false, hasLikedBook = () => false, hasFollowing = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing,}}>
      {children}
    </UserContext.Provider>
);

export { UserProvider, CurrUserProvider, NonUserProvider };
