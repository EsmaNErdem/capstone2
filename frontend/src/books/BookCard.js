import React, { useState } from "react";
import { Link } from 'react-router-dom';
import useBookLike from "../hooks/useBookLike";
import useReviewAdd from "../hooks/useReviewAdd";
import useReviewDelete from "../hooks/useReviewDelete";
import Alert from "../utilities/Alert"
import BookClubApi from "../api";
import BookReviewDrawer from "./BookReviewDrawer";
import IconButton from '@mui/material/IconButton';
import FavoriteIcon from '@mui/icons-material/Favorite';
import CommentIcon from '@mui/icons-material/Comment';
import './BookCard.css'; 

/**
 * Show short info about a book
 * 
 * - Rendered by BookList to display card for each book and shows how many reviews and likes the book has.
 * - Send user like on a book. If fails, shows error to user
 * - Display book review count and like count
 * 
 * - BookList ==> BookCard
 */
const BookCard = ({ id, title, author, description, publisher, category, cover, bookLikeCount=0, reviews=[], likedBook=false, setBooks=null, currUser=false }) => {
    console.debug("BookCard");
        
    const [bookReviews, setBookReviews] = useState(reviews);
    const [isReviewDrawerOpen, setReviewDrawerOpen] = useState(false);
    const [error, setError] = useState(null);

    const { liked, likes, error: likeError, handleLikeBook } = useBookLike(id, bookLikeCount, { id, title, author, description, publisher, category, cover });
    const { error: addError, addBookReview } = useReviewAdd(setBookReviews, { id, title, author, description, publisher, category, cover }, setReviewDrawerOpen)
    const { error: deleteError, deleteBookReview } = useReviewDelete(setBookReviews);

    // open side drawer to show book related reviews
    const openReviewDrawer = async () => {
        try {
            const reviews = await BookClubApi.getAllReviewsByBook(id)
            setBookReviews(reviews)
            setReviewDrawerOpen(true);
        } catch (e) {
            console.error("BookList API call data loading error:", e);
            setError("An error occurred while fetching book reviews.");
        }
    }

    // closes drawer with book related reviews
    const closeReviewDrawer = () => {
        setReviewDrawerOpen(false);
    }

    // sends book likes to database. if BookCard component is called from Profile component, it updates user liked-book list.
    const handleLikeBookButton = () => {
        handleLikeBook();
        if(liked && currUser) {
            setBooks(prevBooks => {
              const updatedBooks = prevBooks.filter(book => book.book_id !== id)
              return updatedBooks
            })
        }
    }

    return (
        <div className="BookCard">
            <Link to={`/books/${id}`} className="Title">
                <img alt={title} src={cover} className="CardImage"/>
            </Link>
            <div className="CardContent">
                <Link to={`/books/${id}`} className="Title" data-testid="book-title-link">
                    <h2 data-testid="book-title">{title}</h2>
                </Link>
                <h3 data-testid="book-author">by {author}</h3>
                <p>{description?.length > 250 ? `${description.slice(0, 250)}...` : description}</p>
                <div className="CardFooter">
                        {!likedBook &&
                        <div style={{ display: 'flex', alignItems: 'center' }}>
                            <span data-testid="review-count">{bookReviews.length}</span>
                            <IconButton  data-testid="review-button" onClick={openReviewDrawer} style={{ color: bookReviews.length > 0 ? "yellowgreen" : ""}}>
                                <CommentIcon  />
                            </IconButton>
                        </div>}
                        <div style={{ display: 'flex', alignItems: 'center' }}>
                            <span data-testid="book-card-like-count">{likes}</span>
                            <IconButton onClick={handleLikeBookButton} color={liked ? 'error' : 'default'}  data-testid="book-card-like-button">
                                <FavoriteIcon/>
                            </IconButton>
                        </div>
                </div>
                {(likeError || addError || deleteError) && <Alert type="danger" messages={[likeError || addError || deleteError]} />}
                {error ? <Alert type="danger" messages={[error]} />: null}
            </div>
            <BookReviewDrawer isOpen={isReviewDrawerOpen} onClose={closeReviewDrawer} reviews={bookReviews} addReviews={addBookReview} deleteReview={deleteBookReview} bookData ={{id, title, author, cover}} />
        </div>
    )
}

export default BookCard;