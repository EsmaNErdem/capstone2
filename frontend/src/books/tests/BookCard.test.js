import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from "../../testUtilities";
import { render, act, fireEvent, waitFor } from '@testing-library/react';
import BookCard from '../BookCard';
import * as useBookLikeModule from '../../hooks/useBookLike';

const book = {
    id: '1',
    title: 'Book1',
    author: 'Author1',
    publisher: 'Publisher1',
    description: 'Description1',
    category: 'Category1',
    cover: 'Cover1',
    bookLikeCount: 2,
    reviews: ["Review1"]
}

jest.mock("../../api", () => ({
    __esModule: true,
    default: {
        getAllReviewsByBook: jest.fn(),
    },
}));

let handleLikeReviewMock;

beforeEach(() => {
  handleLikeBookMock = jest.fn();

  jest.spyOn(useBookLikeModule, 'default').mockImplementation(() => ({
    liked: false,
    likes: 2,
    error: null,
    handleLikeBook: handleLikeBookMock,
  }));
});  

afterEach(() => {
jest.clearAllMocks();
});

test('renders without crashing', async () => {
    await act(async () => {
        render(
            <MemoryRouter>
                <UserProvider>
                    <BookCard />
                </UserProvider>
            </MemoryRouter>
        );
    });

});

test('it renders and matches with snapshot', async () => {
    await act(async () => {
            ({ asFragment } = render(
                <MemoryRouter>
                    <UserProvider>
                        <BookCard
                              id={book.id}
                              title={book.title}
                              author={book.author}
                              description={book.description}
                              publisher={book.publisher}
                              category={book.category}
                              cover={book.cover}
                              bookLikeCount={book.bookLikeCount}
                              reviews={book.reviews} />
                    </UserProvider>
                </MemoryRouter>
            ));
        });
        
    expect(asFragment()).toMatchSnapshot();
});

test('render book data and links to the correct book details page', async () => {
    await act(async () => {
        ({ getByTestId } = render(
            <MemoryRouter>
                <UserProvider>
                    <BookCard
                          id={book.id}
                          title={book.title}
                          author={book.author}
                          description={book.description}
                          publisher={book.publisher}
                          category={book.category}
                          cover={book.cover}
                          bookLikeCount={book.bookLikeCount}
                          reviews={book.reviews} />
                </UserProvider>
            </MemoryRouter>
        ));
    });
    
    expect(getByTestId("book-card-like-count")).toHaveTextContent("2")
    expect(getByTestId("review-count")).toHaveTextContent("1")
    expect(getByTestId("book-title")).toHaveTextContent("Book1")
    const titleLink = getByTestId('book-title-link');
    expect(titleLink.getAttribute('href')).toBe('/books/1');
});

test('implement book like functionality', async () => {
    await act(async () => {
        ({ getByTestId } = render(
            <MemoryRouter>
                <UserProvider>
                    <BookCard
                          id={book.id}
                          title={book.title}
                          author={book.author}
                          description={book.description}
                          publisher={book.publisher}
                          category={book.category}
                          cover={book.cover}
                          bookLikeCount={book.bookLikeCount}
                          reviews={book.reviews} />
                </UserProvider>
            </MemoryRouter>
        ));
    });
    
    const bookLikeButton = getByTestId("book-card-like-button");
    fireEvent.click(bookLikeButton);
    
    await waitFor(() => {
        expect(handleLikeBookMock).toHaveBeenCalled();
    });
});