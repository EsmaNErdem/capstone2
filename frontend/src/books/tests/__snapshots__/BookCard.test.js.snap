// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snapshot 1`] = `
<DocumentFragment>
  <div
    class="BookCard"
  >
    <a
      class="Title"
      href="/books/1"
    >
      <img
        alt="Book1"
        class="CardImage"
        src="Cover1"
      />
    </a>
    <div
      class="CardContent"
    >
      <a
        class="Title"
        data-testid="book-title-link"
        href="/books/1"
      >
        <h2
          data-testid="book-title"
        >
          Book1
        </h2>
      </a>
      <h3
        data-testid="book-author"
      >
        by Author1
      </h3>
      <p>
        Description1
      </p>
      <div
        class="CardFooter"
      >
        <div
          style="display: flex; align-items: center;"
        >
          <span
            data-testid="review-count"
          >
            1
          </span>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-testid="review-button"
            style="color: yellowgreen;"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CommentIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M21.99 4c0-1.1-.89-2-1.99-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4-.01-18zM18 14H6v-2h12v2zm0-3H6V9h12v2zm0-3H6V6h12v2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          style="display: flex; align-items: center;"
        >
          <span
            data-testid="book-card-like-count"
          >
            2
          </span>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-testid="book-card-like-button"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FavoriteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="PrivateSwipeArea-root css-x15wq9"
      style="width: 20px;"
    />
  </div>
</DocumentFragment>
`;
